src/visitor.rs

```diff
@@ -358,209 +396,243 @@ pub trait Visitor<'a> {
+
+            if i < (len - 1) {
+                self.write(", ")?;
+            }
         }
 
-        result.join(", ")
+        Ok(())
     }
 
     /// A visit in the `GROUP BY` section of the query
-    fn visit_grouping(&mut self, grouping: Grouping<'a>) -> String {
-        let mut result = Vec::new();
+    fn visit_grouping(&mut self, grouping: Grouping<'a>) -> fmt::Result {
+        let len = grouping.0.len();
 
-        for value in grouping.0.into_iter() {
-            result.push(self.visit_database_value(value).to_string());
+        for (i, value) in grouping.0.into_iter().enumerate() {
+            self.visit_database_value(value)?;
+
+            if i < (len - 1) {
+                self.write(", ")?;
+            }
         }
 
-        result.join(", ")
+        Ok(())
     }
 
-    fn visit_function(&mut self, fun: Function<'a>) -> String {
-        let mut result = match fun.typ_ {
+    fn visit_function(&mut self, fun: Function<'a>) -> fmt::Result {
+        match fun.typ_ {
             FunctionType::RowNumber(fun_rownum) => {
                 if fun_rownum.over.is_empty() {
-                    String::from("ROW_NUMBER() OVER()")
+                    self.write("ROW_NUMBER() OVER()")?;
                 } else {
-                    format!(
-                        "ROW_NUMBER() OVER({})",
-                        self.visit_partitioning(fun_rownum.over)
-                    )
+                    self.write("ROW_NUMBER() OVER")?;
+                    self.surround_with("(", ")", |ref mut s| s.visit_partitioning(fun_rownum.over))?;
                 }
             }
             FunctionType::Count(fun_count) => {
                 if fun_count.exprs.is_empty() {
-                    String::from("COUNT(*)")
+                    self.write("COUNT(*)")?;
                 } else {
-                    format!("COUNT({})", self.visit_columns(fun_count.exprs))
+                    self.write("COUNT")?;
+                    self.surround_with("(", ")", |ref mut s| s.visit_columns(fun_count.exprs))?;
                 }
             }
             FunctionType::AggregateToString(agg) => {
-                self.visit_aggregate_to_string(agg.value.as_ref().clone())
+                self.visit_aggregate_to_string(agg.value.as_ref().clone())?;
             }
         };
 
         if let Some(alias) = fun.alias {
-            result.push_str(" AS ");
-            result.push_str(&Self::delimited_identifiers(vec![&*alias]));
+            self.write(" AS ")?;
+            self.delimited_identifiers(&[&*alias])?;
         }
 
-        result
+        Ok(())
     }
 
-    fn visit_partitioning(&mut self, over: Over<'a>) -> String {
-        let mut result = Vec::new();
-
+    fn visit_partitioning(&mut self, over: Over<'a>) -> fmt::Result {
         if !over.partitioning.is_empty() {
-            let mut parts = Vec::new();
+            let len = over.partitioning.len();
+            self.write("PARTITION BY ")?;
```

Comment:
not positive, but wouldn't a space at the beginning of the string be safer here?

Reply:
no :(